### ISSUE TITLE
A "pipeline" view of tasks in a job

### ISSUE THREAD

**Issue by pn-santos** [concourse/concourse#2608]
## What challenge are you facing?

In some jobs it can be useful to run alternating "sections" with several parallel tasks and sequential ones. Mix that in with doing regular `put` actions with `on_failure` and `on_success` triggers, a fair amount of logs and the job UI page can very quickly become hard to take in at a glance (e.g. quickly identify which tasks failed when run in parallel).

A practical example: I have a job that builds github PRs and then runs several checks in parallel (using the output of the build). When a check fails it's a chore to scroll up and down the job page to be able to see the logs of the checks that failed.

## What would make this better?

It would be nice if there was a way to see a job the way we see pipelines with horizontal left-to-right layout, no logs, just name of tasks and their status: pass/fail/pending/in progress. This in a way that the tasks relationships were visually represented (i.e. easily identify parallel tasks, if they are `on_<condition>` triggered tasks, etc.).

Then by clicking a task, we'd be able to see only it's log.

I think a nice side effect would be that these screens would be faster to load (no need to load all the logs, just task names, relationships and status) and even when looking at one task, it would only have to load the logs for that task, not the entire job.

## Are you interested in implementing this yourself?

Unfortunately no, this is more of a frontend kind of thing, so not really part of my skill set.

**Comment by jama-pivotal** [comment]
Very interesting! I think it would be a very useful view

**Comment by pn-santos** [comment]
I still think this would be useful, but I have zero chance of making it happen myself.

**Comment by vito** [comment]
Hmm, one change we made was to have all steps auto-collapse by default, even while they're running. Does that help with this? (#4051)

Interestingly there's been some talk about having *running* steps auto-expand/collapse, but that seems to be the root of your issue. /cc @eedwards-sk

One problem with the current approach, aside from whether or not the steps auto-expand, is that it can be a little hard to notice that a step is running. I wonder if we should give them a yellow border? :thinking: /cc @Lindsayauchin @pivotal-jamie-klassen @jomsie

**Comment by pn-santos** [comment]
> Does that help with this?

Assuming logs are lazy loaded, yes that helps. 

That said the "pipeline" view I think would be much better are representing fan-in/fan-out steps in a job imo.

The vertical view can get complicated when there is `parallel/aggregate` + `do` nesting.

**Comment by eedwards-sk** [comment]
@pn-santos hits on one of the biggest things I ran into with concourse and what I commonly see others doing, which is tending to try to shove CI patterns into "pipelines" instead of "jobs" simply because they love the pipeline UI so much

I agree it would be nice if we had what is currently the pipeline UI, but also for jobs. I feel like it would help people stop treating jobs as a "lesser citizen" and trying to do a bunch of weird stuff between jobs that really should be a single job, all because they like watching the glowing boxes.

**Comment by vito** [comment]
Yeah, it does sound interesting. Especially because step names tend to be short and sweet - if everything is auto-collapsed it's kind of a waste of horizontal space.

I'll put this as feedback to consider for our build page redesign. No promises obviously as this would be a pretty big change outside of the initial things we're considering for the redesign.

@eedwards-sk FWIW, one reason to split into separate jobs is also to more easily re-trigger parts that failed. Builds are all-or-nothing.

**Comment by eedwards-sk** [comment]
@vito oh yeah there's tons of reasons for separate jobs! My point is just that I have commonly seen people ending up using pipelines first and trying to do job-behavior with them, only to eventually realize they didn't understand that jobs were more like pipelines in other tools.

The pretty green boxes UI tends to drive people to go, "I want all my steps to be flashing green boxes!"

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
